### DISCUSSION TITLE
Dynamic imports are still including JS on the source code

### DISCUSSION THREAD

**Discussion by VegtMedia-Alex** [vercel/next.js#13668] [Help]
We have a CMS systeem which can contain all kind of blocks. In React we have a component with dynamic imports like this (simplified):

```
const Accordion = dynamic(() => import('components/organisms/Accordion'));
const Alert = dynamic(() => import('components/organisms/Alert'));
const BrandContact = dynamic(() => import('components/organisms/BrandContact'));
const BrandIntro = dynamic(() => import('components/organisms/BrandIntro'));

const config = {
	'accordion': {
		component: Accordion,
	},
	'alert': {
		component: Alert,
	},
	'brandcontact': {
		component: BrandContact,
	},
	'brandintro': {
		component: BrandIntro,
	},
};
```

Then at the render function we are simply doing something like:

```
{blocks && blocks.map((props, index) => {
    let { name, className, marginBottom, columns } = props;
    let block = config[name];
    let Component = block.component;

return (
		<Component />
	);
})
```

I would expect that when a page does **not** use a brandcontact block the js file for this block is not included in the source code on the SSR side.

But at the moment all component js chunks are added to the source code.

Am I doing something wrong or is this a bug?

**Comment by timneutkens** [comment]
We need to load all dynamic imports on the server-side as React does not allow us to suspend rendering, meaning we have to have the underlying module before rendering the React tree. We also don't know what is going to be rendered (because the tree can branch off) hence why all dynamic() loading functions are called before rendering server-side.

Note that only the script tags for dynamic imports that have been rendered will be sent to the browser so this does not affect browser-side hydration.

**Reply by VegtMedia-Alex** [reply]
Hi @timneutkens ,

Thanks for your answer. The browser-side hydration is my problem, at the end of the source code I see a lot chunks being included:

```
<script nomodule="" src="/_next/static/runtime/polyfills-d211fc50b4e12b828420.js"></script>
<script async="" data-next-page="/_app" src="/_next/static/AfVT-YtmIaXPg7XaK7sxo/pages/_app.js"></script>
<script async="" data-next-page="/content" src="/_next/static/AfVT-YtmIaXPg7XaK7sxo/pages/content.js"></script>
<script async="" src="/_next/static/chunks/73.8360c7df0626b75e6bcf.js"></script>
<script async="" src="/_next/static/chunks/a815c4c8ba1a068c6fd82cf76215e54289a0c545.41744ac8b212eca7f083.js"></script>
<script async="" src="/_next/static/chunks/71.882c30b85c67a514f648.js"></script>
<script async="" src="/_next/static/chunks/64.5ffc9e648ce2631d9a8d.js"></script>
<script async="" src="/_next/static/chunks/56.00e595c1836a1093ad5c.js"></script>
<script async="" src="/_next/static/chunks/57.67d26bed9d64f7dc1a56.js"></script>
<script async="" src="/_next/static/chunks/40.ee60fd178a25ab2b0563.js"></script>
<script async="" src="/_next/static/chunks/1c377de9871869784a9a644e97cd16155818224d.f7cada09a4d446ae5271.js"></script>
<script async="" src="/_next/static/chunks/c8f7fe3b0e41be846d5687592cf2018ff6e22687.84f27dc8d0f22bbe7f09.js"></script>
<script async="" src="/_next/static/chunks/74.13a9026c2c0220b71837.js"></script>
<script async="" src="/_next/static/chunks/70.114f6780259b2dba328c.js"></script>
<script async="" src="/_next/static/chunks/61.934e2f9f5c11c9d7eef6.js"></script>
<script async="" src="/_next/static/chunks/59.09da1df3dfd26dae096b.js"></script>
<script src="/_next/static/runtime/webpack-487a6d51819498da9f10.js" async=""></script>
<script src="/_next/static/chunks/framework.e453913a7c1614dfd8d2.js" async=""></script>
<script src="/_next/static/chunks/2edb282b.079e6a9986a985f2feb9.js" async=""></script>
<script src="/_next/static/chunks/commons.5adbeeef4ce372fadccb.js" async=""></script>
<script src="/_next/static/runtime/main-fcf7d88328c774165055.js" async=""></script>
<script src="/_next/static/chunks/bf6678fe.e09500ae47b6bc9314ec.js" async=""></script>
<script src="/_next/static/chunks/afcfaf35.4648d0bc31a70720289b.js" async=""></script>
<script src="/_next/static/chunks/90daeab3.bb0917c3e9e37a8bd604.js" async=""></script>
<script src="/_next/static/chunks/0ff6ff9debf19626318e22cea5eb42191df1164c.539b3f2d38a906f56e2e.js" async=""></script>
<script src="/_next/static/chunks/1087bab2bc5276184d897e8854886927f85139f8.1bdc93fa2fe177d315e6.js" async=""></script>
<script src="/_next/static/chunks/12d35d3acd7b4b3e544b843bea200fa4651517fd.4cf0c81e4758dfce3e84.js" async=""></script>
<script src="/_next/static/chunks/aa5117074756aa5975f53b70091aa7ea343283b7.0a5a2297f17ad8fd3274.js" async=""></script>
<script src="/_next/static/chunks/5d0f862eccf546b97ca05cb57e27cd3bd5f87c38.0687cae0ca74bdfc16e1.js" async=""></script>
<script src="/_next/static/AfVT-YtmIaXPg7XaK7sxo/_buildManifest.js" async=""></script>
<script src="/_next/static/AfVT-YtmIaXPg7XaK7sxo/_ssgManifest.js" async=""></script>
```
And I used @next/bundle-analyzer to see the components for the chunks. I am sure there are a lot chunks included which are not used on the page.

For example the form component is included here and on this particular page there is no form used.

Also when I test using pagespeed I get:

```
Verwijder ongebruikt JavaScript - Remove unused JavaScript
```
<img width="747" alt="Screenshot 2020-06-03 at 10 36 50" src="https://user-images.githubusercontent.com/3672692/83615072-5ea08400-a586-11ea-8720-115330a7687b.png">

Which is also pointing to the form chunk for example. So how can I remove the form chunk JS file from the source code for example?

**Reply by timneutkens** [reply]
That means you must be rendering that particular component causing it to be included in the initial html 🤔

**Reply by VegtMedia-Alex** [reply]
Yes, but that was my original question. It look like the dynamic loaded components are loaded even when they are not called for in the loop?

**Reply by VegtMedia-Alex** [reply]
I am using the following config btw:

```
target: 'experimental-serverless-trace'
```

**Reply by VegtMedia-Alex** [reply]
Do you have any clue? I am debugging this for days now but I really can't find something odd.

**Reply by hysan** [reply]
@alexvandervegt Did you ever figure out what was happening? I was trying something nearly identical to this and was seeing the same issue.

**Reply by VegtMedia-Alex** [reply]
No unfortunately I did not.. If you figure something out, please let me know!

**Comment by revskill10** [comment]
I think you have to include `{ ssr: false }` for lazy dynamic component .
```
const Accordion = dynamic(() => import('components/organisms/Accordion'), {ssr: false});
```

**Reply by VegtMedia-Alex** [reply]
No, this will remove the whole component from the Source Code also when it is used on particular pages.
